Properly add the getter/setter of a VB auto prop to the public API.

CyrusNajmabadi opened the pull request:
Fixes https://github.com/dotnet/roslyn-analyzers/issues/821

CyrusNajmabadi commented:
Tagging @mavasani

CyrusNajmabadi commented:
Note: i could clean up a lot of this logic (especially around needing to pass around bools like 'isImplicitConstructor'.  But i wanted to keep the PR small and focused.  Let me know if you'd prefer a larger cleanup PR, or would prefer this as is.

CyrusNajmabadi commented:
> Does this preserve an exist property contained in the PublicAPI.txt being refactored to an autoproperty without producing diagnostics?

Yes.

AdamSpeight2008 commented:
```vb
Imports System
Namespace NS
  Public Class C
    Private m_P0 As Integer = 1
    Public Property P0() As Integer
      Get
        Return m_P0
      End Get
      Set(value As Integer)
        m_P0 = value
      End Set
    End Property

    Public Property P1() As Integer
  End Class
End Namespace
```
`ShippedAPI.txt`
```
NS.C
NS.C.New() -> Void
NS.C.P0() -> Integer
NS.C.P0(value As Integer) -> Void
```
We get the following diagnostics (mainly the missing auto-property `P1`)
```
Diagnostics:
// Test0.vb(15,21): warning RS0016: Symbol 'implicit get-accessor for P1' is not part of the declared API.
GetBasicResultAt(15, 21, DeclarePublicAPIAnalyzer.DeclareNewApiRule),
// Test0.vb(15,21): warning RS0016: Symbol 'implicit set-accessor for P1' is not part of the declared API.
GetBasicResultAt(15, 21, DeclarePublicAPIAnalyzer.DeclareNewApiRule
```

Let's say we refactor `P0` to be an auto-property say `publicAPI.txt`
```vb
Imports System
Namespace NS
  Public Class C
    Private m_P0 As Integer = 1
    Public Property P0() As Integer
    '  Get
    '    Return m_P0
    '  End Get
    '  Set(value As Integer)
    '    m_P0 = value
    '  End Set
    'End Property

    Public Property P1() As Integer
  End Class
End Namespace
```
produces the following diagnostics.
```
// PublicAPI.Shipped.txt(5,1): warning RS0017: Symbol 'NS.C.P0(value As Integer) -> Void' is part of the declared API, but is either not public or could not be found
GetResultAt(5, 1, DeclarePublicAPIAnalyzer.RemoveDeletedApiRule),
// Test0.vb(6,21): warning RS0016: Symbol 'implicit set-accessor for P0' is not part of the declared API.
GetBasicResultAt(6, 21, DeclarePublicAPIAnalyzer.DeclareNewApiRule),
// Test0.vb(15,21): warning RS0016: Symbol 'implicit get-accessor for P1' is not part of the declared API.
GetBasicResultAt(15, 21, DeclarePublicAPIAnalyzer.DeclareNewApiRule),
// Test0.vb(15,21): warning RS0016: Symbol 'implicit set-accessor for P1' is not part of the declared API.
GetBasicResultAt(15, 21, DeclarePublicAPIAnalyzer.DeclareNewApiRule)
```
In my opinion is incorrect.
  * The existing property `P0` should be considered as being already in the api.
  * It isn't reporting the getter for `P0`
  * Isn't reporting the missing declared public APIs (eg those in the `PublicAPI.txt`)

CyrusNajmabadi commented:
Tests added to validate behavior.

CyrusNajmabadi commented:
@AdamSpeight2008 The main problem here is that the parameter name did change for this accessor.  previously it was 'value', but it changed to 'AutoPropertyValue'.  This is technically a visible change.

AdamSpeight2008 commented:
If we instead use the actual property symbol as the `api name` rather that the getters and setter, then the parameter name doesn't change. But this would report diagnostics for missing apis, for `PublicAPI.txt` that used the "getter and setter" methods. This would be a one time hit, if we use the `property` instead. After that refactor to / from auto-property to full-property would not report diagnostics. Unless there was significant change to property. eg made ReadOnly, a parameter name changes. etc.

CyrusNajmabadi commented:
@AdamSpeight2008 That would be an enormous breaking change to anyone moving to a new version of the public api analyzer.  That would be a non-starter for me.

CyrusNajmabadi commented:
Also, if you'd like to discuss this further, we can do it on gitter.

AdamSpeight2008 commented:
For example
```vbnet.
Imports System
Public Class C
    Private m_Property As Integer
    Public Property [Property]() As Integer
    '   Get
    '      Return m_Property
    '   End Get
    '   Set
    '       m_Property = Value
    '  End Set
    ' End Property
    Public ReadOnly Property ReadOnlyProperty0() As Integer
        Get
            Return m_Property
        End Get
    End Property
    Public WriteOnly Property WriteOnlyProperty0() As Integer
        Set
           m_Property = Value
        End Set
    End Property
    Public ReadOnly Property ReadOnlyProperty1 As Integer = 0
    Public ReadOnly Property ReadOnlyProperty2 As Integer
    Public Property Property1 As Integer
End Class
```
would produce the following `PublicAPI`
```
C
C.New() -> Void
C.Property -> Integer
WriteOnly C.WriteOnlyProperty0 -> Integer
ReadOnly C.ReadOnlyProperty0 -> Integer
ReadOnly C.ReadOnlyProperty1 -> Integer
ReadOnly C.ReadOnlyProperty2 -> Integer
C.Property1 -> Integer
```
Change any of the property to full / auto property wouldn't affect the public APIs.

CyrusNajmabadi commented:
> Change any of the property to full / auto property wouldn't affect the public APIs.

This would be a massive breaking change.  I do not want to go that path.  Thanks.

CyrusNajmabadi commented:
From a strict .net view, these are breaking changes AFAICT.  For example, any .net language that allows calling accessor methods directly, and allows for named parameters, would be affected here if you made that change.  So it comes down to whether or not the PublicApi checker is a general .net checker, or if it is specifically about breaks between VB and C#.  

If the former, the current behavior in my PR is ok.  If the latter, we'd probably want some sort of unification of accessor methods that ignores the method name.

AdamSpeight2008 commented:
The different views are out in the open, so that why I am going to close my pull request. So that an discussion on the way forward.

CyrusNajmabadi commented:
The more i think about it, the more i think the way this PR works is desirable.  Here are the salient points:

1. it includes the getter/setter for the property, something that was missing before, and which made it so that publicapi.txt didn't really represent that these members exist.
2. it makes you consider the implication of what it means to move from a normal VB prop to an auto-prop (this is VB specific as C# does not let you choose setter names).  

Now, for '2', probably all of the time, you won't care, and you'll just update your shipped/unshipped files accordingly to designate the change.  But it will be something you can easily do in the few cases where you make these changes.

mavasani commented:
I haven’t done a thorough review of either PRs, but overall I tend to like Cyrus’s approach much more. I will try to get to this tomorrow. Thanks for the contributions folks!

mavasani commented on src/Roslyn.Diagnostics.Analyzers/UnitTests/DeclarePublicAPIAnalyzerTests.cs:
```diff
@@ -142,7 +142,7 @@ public class C
         }
 
-        [Fact(Skip = "821"), WorkItem(821, "https://github.com/dotnet/roslyn-analyzers/issues/821")]
+        [Fact, WorkItem(821, "https://github.com/dotnet/roslyn-analyzers/issues/821")]
```
Can you also add a unit test that validates the contents of the public API file, probably just port the test added here: https://github.com/dotnet/roslyn-analyzers/pull/2036/files#diff-dd2170e9bef53f186020ef531e877547R206

mavasani commented:
@CyrusNajmabadi The build seems to be failing as you have IDE0055 formatting violations, which are now strictly enforced during build in this repo.

mavasani commented:
@CyrusNajmabadi Would you be able to address the review feedback and fix conflicts? I will merge the PR subsequently. Thanks.

mavasani commented:
Also tagging @genlu, who is driving/owning this analyzer.

genlu commented:
Ping @CyrusNajmabadi 
I will merge this once you have addressed the remaining comments (formatting, test) and resolve the conflicts. Thanks!

CyrusNajmabadi commented:
> Ping @CyrusNajmabadi
I will merge this once you have addressed the remaining comments (formatting, test) and resolve the conflicts. Thanks!

I'm about to go on vacation for a bit.  So maybe you want to take from me and complte this?

genlu commented:
@CyrusNajmabadi Sure. And enjoy the vacation ;)

CyrusNajmabadi commented:
Thanks!
